rds_instance module and tests

s-hertel opened the pull request:
##### SUMMARY
New boto3 rds_instance module that supports encryption and aurora with integration tests.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
rds_instance

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0
```


##### ADDITIONAL INFORMATION
Fixes #19524
Fixes #11159
Fixes #20395
Fixes #22617
Fixes #24415
Fixes #29831
Fixes #24268
Fixes #28981
Fixes #34067
Fixes #38073
Fixes #30020
Fixes #30094
Fixes #34064
Fixes #23618

s-hertel committed:
Add functions to retrieve the allowed and required parameters for boto3 client methods

s-hertel committed:
Add custom waiter for stopping an RDS DB instance

s-hertel committed:
Add rds_instance module

s-hertel committed:
Add rds_instance integration tests

dmsimard commented:
This is very timely -- thanks for you work !
It comes as I was looking at the existing ``rds`` module and noticing it came short on a couple points.

I'll test it out, review and report back.

ryansb commented on test/integration/targets/rds_instance/tasks/main.yml:
```diff
@@ -0,0 +1,12 @@
+- block:
+
+    - include: ./credential_tests.yml
+    - include: ./test_states.yml
```
Having these separated out is lovely.

ryansb commented on test/integration/targets/rds_instance/aliases:
```diff
@@ -0,0 +1,3 @@
+cloud/aws
+shippable/aws/group2
```
If they're unsupported, should they be in a shippable group at all? I have cloudformation_stack_set just grouped as unsupported+cloud/aws

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        storage_encrypted=dict(type='bool'),
+        storage_type=dict(choices=['standard', 'gp2', 'io1']),
+        tags=dict(type='dict'),
+        tde_credential_arn=dict(),
```
I think having this aliased `transparent_data_encryption_*` would be nice. I had to look up what this meant and which DB engines supported it.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        storage_type=dict(choices=['standard', 'gp2', 'io1']),
+        tags=dict(type='dict'),
+        tde_credential_arn=dict(),
+        tde_credential_password=dict(no_log=True),
```
TDE is also only available for MSSQL and Oracle database engines according to https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Overview.Encryption.html - do we want to enforce that, or does the API give a nice error?

dmsimard replied:
Hmmm, I'm not sure the storage_type works with aurora-postgresql. Seeing errors like: ``StorageTypeNotSupportedFault: An error occurred (StorageTypeNotSupported) when calling the CreateDBInstance operation: Invalid storage type: standard`` and ``StorageTypeNotSupportedFault: An error occurred (StorageTypeNotSupported) when calling the CreateDBInstance operation: Invalid storage type: gp2``

dmsimard replied:
Replying to myself, this is an aurora limitation and it's specified in the API docs: https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_CreateDBInstance.html

Maybe we can add in the docstring for ``storage_type`` that it is only for non-aurora instances.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        license_model=dict(choices=['license-included', 'bring-your-own-license', 'general-public-license']),
+        master_user_password=dict(aliases=['password'], no_log=True),
+        master_username=dict(aliases=['username']),
+        monitoring_interval=dict(),
```
I think this should be an `int` since it's always a number of seconds

dmsimard replied:
It actually **needs** to be an int, it otherwise fails like this: ``Invalid type for parameter MonitoringInterval, value: 60, type: <type 'str'>, valid types: <type 'int'>, <type 'long'>"``

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        preferred_maintenance_window=dict(aliases=['maintenance_window']),
+        processor_features=dict(type='dict'),
+        promotion_tier=dict(),
+        publically_accessible=dict(type='bool'),
```
I think `publicly` is the more common spelling for this https://en.wiktionary.org/wiki/publically

s-hertel replied:
Oops. Thanks for catching that. That's even a bug considering it doesn't match the expected CamelCase and won't be applied to any RDS instances.

dmsimard commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+            hyphens and the first character must be a letter and may not end in a hyphen or contain consecutive hyphens.
+        aliases:
+          - cluster_id
+    db_instance_class:
```
I haven't gone through all the options yet but I think it would be great if we could make ``ec2_instance`` and ``rds_instance`` as close as possible in terms of how they are used.
This might translate to changing something here or changing something in ``ec2_instance`` if we feel the implementation here is better.

For example, ``ec2_instance`` has ``instance_type`` for picking the flavor of the instance but here it is ``db_instance_class`` instead.

s-hertel replied:
Thanks so much for reviewing @dmsimard. Let me know if you notice anything else.

My implementation relies on the module options being convertible to the CamelCase boto3 method parameters, so I've been adding aliases rather than modifying the options. Does that work for you? I'll comb through the options for any in common in ec2_instance and add them as aliases. I think modifying the options themselves would make the module harder to maintain (note that this module has 75 options whereas ec2_instance has 35; I want to keep things as simple as possible while being user-friendly).

willthames replied:
If AWS can't be bothered to make RDS and EC2 consistent, we shouldn't strive too hard to do so either :)

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        db_subnet_group_name=dict(),
+        domain=dict(),
+        domain_iam_role_name=dict(),
+        enable_cloudwatch_logs_exports=dict(type='list'),
```
A nice alias would be `cloudwatch_log_exports`

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        tde_credential_arn=dict(),
+        tde_credential_password=dict(no_log=True),
+        timezone=dict(),
+        use_latest_restorable_time=dict(type='bool'),
```
`restore_from_latest` would be a nice alias

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        call_method(
+            client, module, method_name='promote_read_replica_db_instance',
+            parameters={'DBInstanceIdentifier': instance['DBInstanceIdentifier']}
+        )
```
It looks like if an instance is promoted, call_method will attempt to wait if the use sets wait=True, but I don't see a waiter in the RDS docs or in the custom waiters that would cover that. Is that true?

s-hertel replied:
I tested this out (and will add it to the test suite). The db_instance_available built-in waiter is used and works as expected https://boto3.readthedocs.io/en/latest/reference/services/rds.html#RDS.Waiter.DBInstanceAvailable.

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+            if it needs to be updated is not possible this is set to False by default to allow idempotence.
+        type: bool
+        default: False
+    promote:
```
while you've implemented `promote` in a declarative fashion, the parameter name doesn't seem declarative.

I wonder if there's a better way to do this. In my original attempt at rds_instance, I used the fact a db was a replication_source, when passed without a source_db_identifier, to promote, but that seems far too dangerous in hindsight.

`read_replica: False` might be better, with `read_replica: True` being implied if `source_db_identifier` is passed.

s-hertel replied:
That sounds much better, thanks for the suggestion. I will update that.

dmsimard commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        db_instance_identifier=dict(required=True, aliases=['instance_id', 'id']),
+        db_name=dict(),
+        db_parameter_group_name=dict(),
+        db_security_groups=dict(),
```
This needs to be type='list' -- the doc mentions it's a list but it's not forced as such which can yield the error ``Invalid type for parameter DBSecurityGroups, value: sg-1234567890, type: <type 'str'>, valid types: <type 'list'>, <type 'tuple'>``

dmsimard commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+    vpc_security_group_ids:
+        description:
+          - A list of EC2 VPC security groups to associate with the DB cluster.
+
```
Missing ``type: list``

dmsimard commented:
There seems to be an argument missing for supplying the VPC ID ?
I'm trying to provision an aurora-postgresql in a specific vpc and the only vpc parameter I've got is ``vpc_security_group_ids``.

If I use that argument, I get the following error:
```
ClientError: An error occurred (InvalidParameterCombination) when calling the CreateDBInstance operation: The DB instance and EC2 security group are in different VPCs. The DB instance is in vpc-aeeecaca and the EC2 security group is in vpc-0ad130989857bcf4e
```

dmsimard commented:
Doh: ``"Unable to create a new DB instance: An error occurred (InvalidParameterCombination) when calling the CreateDBInstance operation: Standalone DB Instances are not supported for the specified engine"`` I suppose that aurora-postgresql would require using the (missing) ``rds_cluster`` module.

willthames commented:
@dmsimard - for Aurora you'd need to create the cluster first (using #42128 ?) and then pass cluster id to the rds_instance module. Can't quite work out the storage_type settings - not sure whether there just isn't a choice for Aurora or there's a limited choice

s-hertel commented:
@willthames I don’t think storage type is accepted for Aurora instances. I don’t see where AWS has this documented, but IIRC Aurora DB instances use SSD-backed virtualized storage layer rather than EBS volumes like the other engines.

dmsimard commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1236 @@
+        master_username=dict(aliases=['username']),
+        monitoring_interval=dict(),
+        monitoring_role_arn=dict(),
+        multi_az=dict(type='bool'),
```
Looking at the docs (https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_CreateDBInstance.html) I think multi_az should be mutually exclusive with zones. ``Specifies if the DB instance is a Multi-AZ deployment. You can't set the AvailabilityZone parameter if the MultiAZ parameter is set to true. ``

s-hertel replied:
The availability_zones parameter is a little broken at the moment. I'll make them mutually exclusive when I get that working.

dmsimard replied:
Right, setting multi_az to true doesn't seem to be doing anything special currently (my RDS postgre instance ends up being created without multi-az enabled)

s-hertel committed:
address requested changes from ryansb

s-hertel committed:
address requested changes from willthames

s-hertel commented:
@dmsimard Do you happen to have the task at hand that resulted in this or have the full traceback? https://github.com/ansible/ansible/pull/43789#issuecomment-411274405

I want to direct people to rds_cluster with clearer errors for things like this (you can add aurora* instances to pre-existing clusters but not create a standalone instance). I tried to reproduce by omitting the cluster identifier and using a nonexistent cluster. Without providing cluster_id I get the error:
```
"msg": "engine is aurora but all of the following are missing: cluster_id"
```
And with a nonexistent cluster ID I get this:
```
"msg": "Unable to create a new DB instance: An error occurred (DBClusterNotFoundFault) when calling the CreateDBInstance operation: Could not find DB Cluster: doesnotexist"
```

s-hertel committed:
address requested changes from dmsimard

s-hertel committed:
Fix final snapshots

Fix idempotence with already-deleting DB instances

Remove unused import from module_utils/aws/core.py

Consolidate function to get all boto3 client method parameters and the subset of required parameters

s-hertel commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1267 @@
+        description:
+          - The name of the DB parameter group to associate with this DB instance. When creating the DB instance if this
+            argument is omitted the default DBParameterGroup for the specified engine is used.
+    db_security_groups:
```
@dmsimard To make this more like ec2_instance I was going to consolidate `db_security_groups` and `vpc_security_group_ids` into one `security_groups` option (retaining the original option names as aliases), but decided not to since I don't have access to an EC2-Classic account and cannot test the edge cases.

dmsimard replied:
Yeah I think it's fair to keep the module consistent with the API (despite being inconsistent with other AWS modules, it's not technically out fault) for the sake of keeping things maintainable. It's a shame from a UX perspective though.

s-hertel committed:
Add some additional rds_instance integration tests

s-hertel committed:
Add some common functions to module_utils/aws/rds

s-hertel committed:
Move common code out of rds_instance

ryansb commented on lib/ansible/module_utils/aws/rds.py:
```diff
@@ -0,0 +1,210 @@
+
+def wait_for_status(client, module, identifier, method_name):
+    is_cluster = get_rds_method_attribute(method_name, module).cluster
+    is_instance = get_rds_method_attribute(method_name, module).instance
```
These vars are never used

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1152 @@
+          - Set to False to promote a read replica cluster or true to create one. When creating a read replica C(creation_source) should
+            be set to 'instance' or not provided. C(source_db_instance_identifier) must be provided with this option.
+        type: bool
+    wait:
```
it doesn't look like there's an accompanying wait_timeout option. Can we add to the docs what the max time the module will wait, or if it can hang indefinitely if RDS hangs?

s-hertel replied:
I added to the docs and made the made the possible wait time less ridiculously long than I had it.

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1152 @@
+        type: bool
+    engine:
+        description:
+          - The name of the database engine to be used for this DB instance. This is required to create an instance.
```
Is there somewhere in boto3 we can pull this out from? It would be great to not have to update this if Amazon adds more engine options.

s-hertel replied:
I could pull it from https://github.com/boto/botocore/blob/c632931a6cc5eab4113c976d430bcb9c059f829f/botocore/data/rds/2014-09-01/service-2.json#L3045? It doesn't look like boto3 or botocore ever validate or do anything to the engine other than specify it to be a string. They just pass along the value. We could do that too. But then the user won't know what the valid values are without looking it up. I would be okay with removing `choices` from this option in the argspec (I'd have to move the documented choices to the description) and add the list as accepted (last known and in need of occassional-updating) enginges to the error message in case it is a bad value. That wouldn't impede unattended growth. Sound okay?

ryansb replied:
Yeah, maybe you could add to the documentation on the option the current engines, but remove the `choices` validation so that new engines will still work but may be undocumented for up to 1 release.

s-hertel replied:
Addressed, also added a helpful error for invalid engines

ryansb commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1152 @@
+        required_if=required_if,
+        mutually_exclusive=mutually_exclusive,
+        supports_check_mode=True
+    )
```
I'm thinking this would be a good one to set a `boto3_at_least` of 1.5.0, which looks like when the TDE options were added. It was also released about one year ago, so would match with the boto3 support policy we'd had in mind.

s-hertel replied:
Addressed

s-hertel committed:
Remove hardcoded engine choices and require the minimum boto3

s-hertel committed:
Document wait behavior

s-hertel committed:
Provide a list of valid engines in the error message if it is invalid

Add supported methods to whitelist

Remove AWSRetry around waiter

Wait for a less crazy amount of time

Remove unused variables

s-hertel committed:
Add a test for an invalid engine option

s-hertel committed:
pep8

ryansb commented:
There seems to be an issue with state=rebooted, if you choose an instance by ID: 

```
- name: try to reboot
  rds_instance:
    state: rebooted
    id: my-favorite-db
```

Results in:

```
    Traceback (most recent call last):                                                                                                                                                                              
      File "/home/ryansb/.ansible/tmp/ansible-tmp-1535401606.73-107835130156339/AnsiballZ_rds_instance.py", line 113, in <module>                                                                                   
        _ansiballz_main()                                                                                                                                                                                           
      File "/home/ryansb/.ansible/tmp/ansible-tmp-1535401606.73-107835130156339/AnsiballZ_rds_instance.py", line 105, in _ansiballz_main                                                                            
        invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)                                                                                                                                                      
      File "/home/ryansb/.ansible/tmp/ansible-tmp-1535401606.73-107835130156339/AnsiballZ_rds_instance.py", line 48, in invoke_module                                                                               
        imp.load_module('__main__', mod, module, MOD_DESC)
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/__main__.py", line 1132, in <module>
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/__main__.py", line 1113, in main
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/__main__.py", line 933, in update_instance
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/__main__.py", line 955, in update_instance_state                                                                                                       
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/__main__.py", line 967, in reboot_running_db_instance                                                                                                  
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/ansible_rds_instance_payload.zip/ansible/module_utils/aws/rds.py", line 119, in call_method                                                            
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/ansible_rds_instance_payload.zip/ansible/module_utils/aws/rds.py", line 59, in get_final_identifier                                                    
      File "/tmp/ryansb/ansible_rds_instance_payload_147z4_/ansible_rds_instance_payload.zip/ansible/module_utils/aws/rds.py", line 52, in get_rds_method_attribute                                                
    NotImplementedError: method reboot_db_instance hasn't been added to the list of accepted methods to use a waiter in module_utils/aws/rds.py         
```

s-hertel committed:
Missed adding a method to the whitelist

s-hertel committed:
Use retries

ryansb commented:
When making a long-running change, like instance size, if you make the same change successively like this:

```
---
- hosts: localhost
  module_defaults:
    rds_instance:
      profile: slscode
      region: us-east-2
    group/aws:
      profile: slscode
      region: us-east-2
  tasks:
  - rds_subnet_group:
      name: test-sub-group
      state: present
      description: fake-group
      subnets:
        - subnet-08370cc6c83f0d865
        - subnet-0761385c754f09bb7
  - rds_instance:
      id: rsb-test
      db_name: rsbio
      engine: postgres
      instance_type: db.t2.micro
      db_subnet_group_name: test-sub-group
      password: 123456789
      username: rsb
      allocated_storage: 10
      multi_az: true
      apply_immediately: true
      wait: true
      vpc_security_group_ids:
        - sg-00a8629192b8df910

  - pause: minutes=3

  - rds_instance:
      id: rsb-test
      db_name: rsbio
      engine: postgres
      instance_type: db.t2.small
      db_subnet_group_name: test-sub-group
      password: 123456789
      username: rsb
      allocated_storage: 10
      multi_az: true
      apply_immediately: true
      wait: true
      vpc_security_group_ids:
        - sg-00a8629192b8df910

  - rds_instance:
      id: rsb-test
      db_name: rsbio
      engine: postgres
      instance_type: db.t2.small
      db_subnet_group_name: test-sub-group
      password: 123456789
      username: rsb
      allocated_storage: 10
      multi_az: true
      apply_immediately: true
      wait: true
      vpc_security_group_ids:
        - sg-00a8629192b8df910
```

The second task with `db.t2.small` instance size will fail because the modification from the first task was `apply_immediately: true` and so the instance is in the `modifying` state, causing the module to fail.

s-hertel committed:
Fix some little things

s-hertel committed:
Fix more things

s-hertel committed:
Improve error message

s-hertel committed:
Support creating cross-region read replicas

ryansb committed:
Remove unused imports

ryansb committed:
Add retry when getting RDS instance

ryansb committed:
Soft-check required options so module fails properly when options are missing

ryansb committed:
Fix mariadb parameter version

s-hertel committed:
Fix cross-region read_replica creation and tests

s-hertel committed:
fix modify tests

willthames commented on lib/ansible/modules/cloud/amazon/rds_instance.py:
```diff
@@ -0,0 +1,1157 @@
+    )
+
+    if not module.boto3_at_least('1.5.0'):
+        module.fail_json(msg="rds_intance requires boto3 > 1.5.0")
```
minor typo

s-hertel committed:
Fix a modification test

s-hertel committed:
Fix typo

s-hertel committed:
Remove test for option_group_name that exists for this account but may not for others and added as a TODO to do properly

s-hertel commented:
bot_status

willthames commented:
Merged! Woohoo! At long last! Well done @s-hertel!
